sktime/annotation/e_agglo.py

```diff
@@ -0,0 +1,435 @@
+        _member = np.array(
+            self.member if self.member is not None else range(X.shape[0])
+        )
+
+        u = np.sort(np.unique(_member))  # unique array of cluster labels
+        n_cluster = len(u)  # number of clusters
+
+        for i in range(
+            n_cluster
+        ):  # relabel clusters to be consecutive numbers (when user specified)
+            _member[np.where(_member == u[i])[0]] = i
+
+        # check if sorted
+        assert all(sorted(_member) == _member)
+
+        sizes = np.repeat(0, 2 * n_cluster)
+        sizes[:n_cluster] = [
+            sum(_member == i) for i in range(n_cluster)
+        ]  # calculate initial cluster sizes
+
+        # array of within distances
+        XX = X.copy().set_index(_member)
+        g = XX.groupby(level=0)
+        within = g.apply(lambda x: get_distance(x, x, self.alpha))
+
+        # array of between between-within distances
+        distances = np.empty((2 * n_cluster, 2 * n_cluster))
+
+        for i, xi in g:
+            distances[:n_cluster, i] = (
+                2 * g.apply(lambda xj: get_distance(xi, xj, self.alpha))  # noqa
+                - within[i]
+                - within
+            )
+
+        np.fill_diagonal(distances, 0)
+
+        # set up left and right neighbors
+        # special case for clusters 0 and n_cluster-1 to allow for cyclic merging
+        left = np.repeat(0, 2 * n_cluster - 1)
+        left[:n_cluster] = [
+            i - 1 if i - 1 >= 0 else n_cluster - 1 for i in range(n_cluster)
+        ]
+        right = np.repeat(0, 2 * n_cluster - 1)
+        right[:n_cluster] = [
+            i + 1 if i + 1 < n_cluster else 0 for i in range(n_cluster)
+        ]
+
+        # True means that a cluster has not been merged
+        open = np.array([True for _ in range(2 * n_cluster - 1)])
+
+        # which clusters were merged at each step
+        merged = np.empty((n_cluster - 1, 2))
+
+        # set initial GOF value
+        gof = np.array(
+            [
+                sum(
+                    [
+                        distances[i, left[i]] + distances[i, right[i]]
+                        for i in range(n_cluster)
+                    ]
+                )
+            ]
+        )
+
+        # change point progression
+        progression = np.empty((n_cluster, n_cluster + 1))
+        progression[0, :] = [
+            sum(sizes[:i]) if i > 0 else 0 for i in range(n_cluster + 1)
+        ]  # N + 1 for cyclic mergers
+
+        # array to specify the starting point of a cluster
+        lm = np.repeat(0, 2 * n_cluster - 1)
+        lm[:n_cluster] = range(n_cluster)
+
+        # store to self
+        self._member = _member
+        self.n_cluster = n_cluster
+        self.sizes = sizes
+        self.distances = distances
+        self.left = left
+        self.right = right
+        self.open = open
+        self.merged = merged
+        self.gof = gof
+        self.progression = progression
+        self.lm = lm
```

Comment:
It's probably less of a gotcha if those are assigned directly where they are computed.